### PR TITLE
JSUI-2467 Populate GTM datalayer with Coveo UA event data

### PR DESCRIPTION
**Pull request by fbeaudoincoveo**
[Deploy](https://dashboard.heroku.com/pipelines/a3535101-5bbf-4a5b-a909-47fcf8c9f149)

### CODE REVIEW

File: src/ui/Analytics/Analytics.ts [hunk 1]
```diff
@@ -463,6 +469,14 @@ export class Analytics extends Component {
     );
   }
 
+  private handleCoveoUAEventReady(data: ICoveoUAEventArgs) {
+    if (this.options.populateDataLayer) {
```
Comment (samisayegh) [inline src/ui/Analytics/Analytics.ts]: I would invert the conditions to have more defensive and less nested control-flow. e.g.
```
private handleCoveoUAEventReady(data: ICoveoUAEventArgs) {
    if (!this.options.populateDataLayer) {
        return;
    }

    const dataLayerName = this.options.dataLayerName;

    if (!<any>window)[dataLayerName]) {
        (<any>window)[dataLayerName] = []
    }

    (<any>window)[dataLayerName].push(data);
}
```

File: src/ui/Analytics/Analytics.ts [hunk 2]
```diff
@@ -463,6 +469,14 @@ export class Analytics extends Component {
...
+      let dataLayerName = this.options.dataLayerName;
```
Comment (samisayegh) [inline src/ui/Analytics/Analytics.ts]: It's a best-practice to use `const` over `let` when the variable value does not change after initialization.

File: src/ui/Analytics/LiveAnalyticsClient.ts [hunk 3]
```diff
@@ -174,8 +175,13 @@ export class LiveAnalyticsClient implements IAnalyticsClient {
     var customEvent = this.buildCustomEvent(actionCause, metaObject, element);
     this.triggerChangeAnalyticsCustomData('CustomEvent', metaObject, customEvent);
     this.checkToSendAnyPendingSearchAsYouType(actionCause);
+    var convertedCustomEvent = APIAnalyticsBuilder.convertCustomEventToAPI(customEvent);
```
Comment (samisayegh) [inline src/ui/Analytics/LiveAnalyticsClient.ts]: It would be good to use `const` here. Same for the `var` on line 175 and 251.

File: src/ui/Analytics/LiveAnalyticsClient.ts [hunk 4]
```diff
@@ -174,8 +175,13 @@ export class LiveAnalyticsClient implements IAnalyticsClient {
...
     $$(this.rootElement).trigger(AnalyticsEvents.customEvent, <IAnalyticsCustomEventArgs>{
-      customEvent: APIAnalyticsBuilder.convertCustomEventToAPI(customEvent)
+      customEvent: convertedCustomEvent
+    });
+    $$(this.rootElement).trigger(AnalyticsEvents.coveoUAEventReady, <ICoveoUAEventArgs>{
```
Comment (samisayegh) [inline src/ui/Analytics/LiveAnalyticsClient.ts]: I left a similar comment on another PR related to using `UA` in the name. I just want to make sure the naming is consistent within the scope of the project. Do we normally use `UA` or `Analytics` or something else?

I find it's a little tidier when the naming is consistent :)

**Comment by samisayegh** [comment]
Small comment for the future; just including the project acronym and jira number in the branch name is perfect. In this case it would be `JSUI-2467`.

Thanks for your contribution 💯

File: src/ui/Analytics/Analytics.ts [hunk 5]
```diff
@@ -158,7 +161,9 @@ export class Analytics extends Component {
      * Default value is `undefined`, and the value of this parameter will fallback to the organization used for the
      * search endpoint.
      */
-    organization: ComponentOptions.buildStringOption()
+    organization: ComponentOptions.buildStringOption(),
+    populateDataLayer: ComponentOptions.buildBooleanOption({ defaultValue: true }),
```
Comment (olamothe) [inline src/ui/Analytics/Analytics.ts]: I'm not certain the name "populateDataLayout" is clear enough ?

How about having some kind of naming that explicitly mention google tag manager ?

Would also be interesting to see the accompanying documentation for that option

File: src/ui/Analytics/Analytics.ts [hunk 6]
```diff
@@ -463,6 +469,14 @@ export class Analytics extends Component {
     );
   }
 
+  private handleCoveoUAEventReady(data: ICoveoUAEventArgs) {
+    if (this.options.populateDataLayer) {
+      let dataLayerName = this.options.dataLayerName;
+      (<any>window)[dataLayerName] = (<any>window)[dataLayerName] || [];
```
Comment (olamothe) [inline src/ui/Analytics/Analytics.ts]: You can augment the global window interface to declare that it contains a property for `dataLayerName` instead of having to cast the global window object to any.

https://stackoverflow.com/a/12709880



Is that global variable declared on "window" really changing and different for each website using google tag manager ? 

Could we set a good default value for the option ?

We should also handle that the option might be misconfigured, and that `window[dataLayerName] == null`, or that `window[dataLayerName]` is somehow something totally different that google tag manager (and that the `push` function would not be available).

File: src/events/AnalyticsEvents.ts [hunk 7]
```diff
@@ -21,6 +21,11 @@ export interface IAnalyticsCustomEventArgs {
   customEvent: IAPICustomEvent;
 }
 
+export interface ICoveoUAEventArgs {
+  event: string;
```
Comment (olamothe) [inline src/events/AnalyticsEvents.ts]: The `event` could be typed to `'CoveoCustomEvent' | 'CoveoClickEvent' | 'CoveoSearchEvent'` instead of generic string

File: src/ui/Analytics/Analytics.ts [hunk 8]
```diff
@@ -158,7 +161,9 @@ export class Analytics extends Component {
      * Default value is `undefined`, and the value of this parameter will fallback to the organization used for the
      * search endpoint.
      */
-    organization: ComponentOptions.buildStringOption()
+    organization: ComponentOptions.buildStringOption(),
+    populateDataLayer: ComponentOptions.buildBooleanOption({ defaultValue: true }),
+    dataLayerName: ComponentOptions.buildStringOption({ defaultValue: 'dataLayer' })
```
[the remark this answers is not in the document]
Comment (fbeaudoincoveo) [reply src/ui/Analytics/Analytics.ts]: Yes adding `depends` makes perfect sense

File: src/ui/Analytics/Analytics.ts (hunk 1 above)
Comment (fbeaudoincoveo) [reply src/ui/Analytics/Analytics.ts]: 👍

File: src/ui/Analytics/Analytics.ts (hunk 2 above)
Comment (fbeaudoincoveo) [reply src/ui/Analytics/Analytics.ts]: Thanks, I'll remember this :)

File: src/ui/Analytics/LiveAnalyticsClient.ts (hunk 4 above)
Comment (fbeaudoincoveo) [reply src/ui/Analytics/LiveAnalyticsClient.ts]: I think we typically use `Analytics`.

I'll also remove the `Coveo` prefix

Comment (fbeaudoincoveo) [reply src/ui/Analytics/LiveAnalyticsClient.ts]: (I'll keep the `coveo` prefix for the IAnalyticsEventArgs `coveoAnalyticsEventData` property however, since it populates the dataLayer, we'll want to make sure it's properly scoped to reduce the odds of it overriding other stuff).

File: src/ui/Analytics/Analytics.ts (hunk 5 above)
Comment (fbeaudoincoveo) [reply src/ui/Analytics/Analytics.ts]: I'll go with:

`pushToGtmDataLayer` and `gtmDataLayerName`

Documentation incoming :)

File: src/ui/Analytics/Analytics.ts (hunk 6 above)
Comment (fbeaudoincoveo) [reply src/ui/Analytics/Analytics.ts]: I wouldn't augment the window interface in this case, because we can't be a 100% sure the property is going to be named `dataLayer`.

It is possible for users to rename the dataLayer in each website, though I doubt this is common practice. The current default (`dataLayer`) is also the name that is output by default when you generate GTM code snippets, so I think that's a good default value.

I'm adding some logic to ensure that the dataLayer is an object, and that it contains at least an item with the `gtm.uniqueEventId` property before pushing anything.

I'm also adding logic to assert that a GTM code snippet is present in the page before even registering the event handler.

Comment (fbeaudoincoveo) [reply src/ui/Analytics/Analytics.ts]: (about renaming the dataLayer: https://developers.google.com/gtagjs/devguide/rename)

File: src/events/AnalyticsEvents.ts (hunk 7 above)
Comment (fbeaudoincoveo) [reply src/events/AnalyticsEvents.ts]: 👍

File: src/ui/Analytics/Analytics.ts [hunk 9]
```diff
@@ -463,6 +492,31 @@ export class Analytics extends Component {
     );
   }
 
+  private isGtmScriptPresent() {
```
Comment (fbeaudoincoveo) [inline src/ui/Analytics/Analytics.ts]: I added this so that even if we do end up setting `pushToGtmDataLayer` to `true` by default (which is currently the case), `handleAnalyticsEventReady` will typically only be called if a GTM container snippet is present in the page.

I made the test none too restrictive, as a GTM snippet can take various forms such as:

- https://stackoverflow.com/questions/47457797/rename-the-datalayer-parameter-for-the-new-gtag-js
- https://developers.google.com/tag-manager/quickstart
or

**Comment by fbeaudoincoveo** [comment]
Concerning unit tests, here's what I have in mind:

`Analytics`:
with default configuration:
- Should push to dataLayer if page contains GTM snippet
- Should not push to dataLayer if page does not contain GTM snippet
- Should not attempt to push to dataLayer if dataLayer.push isn't available

with `gtmDataLayerName` set to [name] other than default (i.e., `dataLayer`):
- Should push to [name] if page contains GTM snippet
- Should not push to [name] if page does not contain GTM snippet
- Should not attempt to push to [name] if dataLayer.push isn't available

with `pushToGtmDataLayer` set to `false`
- Should not push to dataLayer even if page contains GTM snippet

`LiveAnalyticsClient`:
- Should trigger analyticsEventReady on each click event
- Should trigger analyticsEventReady on each custom event
- Should trigger analyticsEventReady on each distinct search event

Anything more I should cover?

**Comment by samisayegh** [comment]
> Concerning unit tests, here's what I have in mind:
> 
> `Analytics`:
> with default configuration:
> 
> * Should push to dataLayer if page contains GTM snippet
> * Should not push to dataLayer if page does not contain GTM snippet
> * Should not attempt to push to dataLayer if dataLayer.push isn't available
> 
> with `gtmDataLayerName` set to [name] other than default (i.e., `dataLayer`):
> 
> * Should push to [name] if page contains GTM snippet
> * Should not push to [name] if page does not contain GTM snippet
> * Should not attempt to push to [name] if dataLayer.push isn't available
> 
> with `pushToGtmDataLayer` set to `false`
> 
> * Should not push to dataLayer even if page contains GTM snippet
> 
> `LiveAnalyticsClient`:
> 
> * Should trigger analyticsEventReady on each click event
> * Should trigger analyticsEventReady on each custom event
> * Should trigger analyticsEventReady on each distinct search event
> 
> Anything more I should cover?

That sounds good

File: src/ui/Analytics/Analytics.ts [hunk 10]
```diff
@@ -199,6 +225,9 @@ export class Analytics extends Component {
 
     this.bind.onRootElement(QueryEvents.buildingQuery, (data: IBuildingQueryEventArgs) => this.handleBuildingQuery(data));
     this.bind.onRootElement(QueryEvents.queryError, (data: IQueryErrorEventArgs) => this.handleQueryError(data));
+    if (this.isGtmScriptPresent()) {
```
Comment (olamothe) [inline src/ui/Analytics/Analytics.ts]: As discussed live, I think this should be dictated solely by the `pushToGtmDataLayer` option.

I'd keep this logic that tries to determine the presence or absence of google tag manager to determine what the "default" value for the option should be, in the case that it's left undefined.

File: src/ui/Analytics/Analytics.ts [hunk 11]
```diff
@@ -463,6 +492,31 @@ export class Analytics extends Component {
...
+  }
+
+  private handleAnalyticsEventReady(data: IAnalyticsEventArgs) {
+    if (!this.options.pushToGtmDataLayer) {
+      return;
+    }
+
+    const dataLayerName = this.options.gtmDataLayerName;
+
+    if (!(<any>window)[dataLayerName]) {
```
Comment (olamothe) [inline src/ui/Analytics/Analytics.ts]: You could declare 
```
const win = window as any

[... rest of the function ...]
```

to prevent the repetion

File: src/ui/Analytics/Analytics.ts [hunk 12]
```diff
@@ -463,6 +492,31 @@ export class Analytics extends Component {
...
+      (<any>window)[dataLayerName] = [];
+    } else if (
+      typeof (<any>window)[dataLayerName] === 'object' &&
+      typeof (<any>window)[dataLayerName].push === 'function' &&
+      (<any>window)[dataLayerName].push !== Array.prototype.push
```
Comment (olamothe) [inline src/ui/Analytics/Analytics.ts]: I think only checking for the presence of `window[dataLayerName]` should be sufficient.

And I'd wrap the call to `window[dataLayerName].push` to a try catch, which simply logs the error in an explicit manner in the console, but does not block the rest of the execution (eg: "normal" coveo analytics call)

File: src/ui/Analytics/Analytics.ts [hunk 13]
```diff
@@ -158,7 +162,69 @@ export class Analytics extends Component {
...
+     *
+     * **Notes:**
+     * - Set this option to `false` to ensure that the component does _not_ populate the data layer, even if a GTM container snippet is present in the page.
+     * - Set this option to `true` to ensure that the component populates the data layer even if a GTM container snippet cannot be found in the page.
+     *
+     * See also [`gtmDataLayerName`]{@link Analytics.options.gtmDataLayerName}.
+     *
+     * By default, if a standard GTM container snippet is found in the page, the component automatically attempts to push to its initialized data layer (see [Google Tag Manager Quick Start Guide](https://developers.google.com/tag-manager/quickstart)).
+     */
+    pushToGtmDataLayer: ComponentOptions.buildBooleanOption({
```
Comment (fbeaudoincoveo) [inline src/ui/Analytics/Analytics.ts]: Essentially, the default behavior is:

- If no GTM container script is found in the page, do nothing.
- If a GTM container snippet is found, set `pushToGtmDataLayer` to `true`, and attempt to extract the data layer name from the GTM container snippet and set `gtmDataLayerName` to this value (set `gtmDataLayerName` to `"dataLayer"` if no data layer name can be extracted from the snippet). If the data layer object is undefined, don't register the handler (to avoid repeatedly trying to push to a non-existing object).

File: src/ui/Analytics/Analytics.ts [hunk 14]
```diff
@@ -158,7 +162,69 @@ export class Analytics extends Component {
...
+          const initializesGtmScript = new RegExp(`\(window,document,'script','(\\w+)','GTM-[A-Z0-9]{7}'\)`);
+          const gtmContainerSnippet = _.find(document.querySelectorAll('script'), scriptElement => {
+            // Could yield false positives (e.g., if the innerText is commented out).
+            return includesGtmScript.test(scriptElement.innerText) && initializesGtmScript.test(scriptElement.innerText);
+          });
+          if (gtmContainerSnippet) {
+            const match = initializesGtmScript.exec(gtmContainerSnippet.innerText);
+            // Attempt to extract data layer name from GTM container snippet
+            if (match.length > 2) {
+              options.gtmDataLayerName = match[2];
```
Comment (fbeaudoincoveo) [inline src/ui/Analytics/Analytics.ts]: I would have liked to use a named group in the regular expression and refer to this, rather than using the index position to access the string I need in the match array. However, I can't seem to find a way to do this. Any ideas?

File: src/ui/Analytics/Analytics.ts [hunk 15]
```diff
@@ -158,7 +162,69 @@ export class Analytics extends Component {
...
+        } else {
+          const includesGtmScript = new RegExp(`https:\/\/www\.googletagmanager\.com\/gtm\.js`);
+          const initializesGtmScript = new RegExp(`\(window,document,'script','(\\w+)','GTM-[A-Z0-9]{7}'\)`);
+          const gtmContainerSnippet = _.find(document.querySelectorAll('script'), scriptElement => {
+            // Could yield false positives (e.g., if the innerText is commented out).
+            return includesGtmScript.test(scriptElement.innerText) && initializesGtmScript.test(scriptElement.innerText);
+          });
+          if (gtmContainerSnippet) {
+            const match = initializesGtmScript.exec(gtmContainerSnippet.innerText);
+            // Attempt to extract data layer name from GTM container snippet
```
Comment (fbeaudoincoveo) [inline src/ui/Analytics/Analytics.ts]: It would arguably be cleaner to do this in the `gtmDataLayerName` option post processing, but this would require the regexes to be tested again (worst case, against the inner text of all script tags the page).

Doing this here also means that if `pushToGtmDataLayer` is left undefined and the `gtmDataLayerName` value is explicitly specified in the component configuration, the specified value will always be overridden if a GTM container snippet is found in the page. My intuition is this would typically "fix" a configuration mistake, but it would also make sense if an explicitly specified option value always took precedence over an automatically generated one.

Thoughts?

File: src/ui/Analytics/Analytics.ts (hunk 14 above)
Comment (samisayegh) [reply src/ui/Analytics/Analytics.ts]: If you just want the original string, would it be possible to set:
`options.gtmDataLayerName = gtmContainerSnippet.innerText`

Or does `match[2]` hold a different value?

File: src/ui/Analytics/Analytics.ts (hunk 13 above)
Comment (samisayegh) [reply src/ui/Analytics/Analytics.ts]: Is the idea that someone who has a GTM script in their page can use this option to tell the search-ui framework not to push to it?

If so, setting pushToGtmDataLayer to `true` automatically when the script is there does not make sense to me.

Regarding extracting the data-layer name, would it be simpler to ask the user to configure it? Maybe this option could become `gtmDataLayerName`? If it is specified, we push to it?

What do you think of having a slack call with share screen to discuss?

File: src/ui/Analytics/Analytics.ts [hunk 16]
```diff
@@ -158,7 +162,69 @@ export class Analytics extends Component {
...
+     * - Set this option to `false` to ensure that the component does _not_ populate the data layer, even if a GTM container snippet is present in the page.
+     * - Set this option to `true` to ensure that the component populates the data layer even if a GTM container snippet cannot be found in the page.
+     *
+     * See also [`gtmDataLayerName`]{@link Analytics.options.gtmDataLayerName}.
+     *
+     * By default, if a standard GTM container snippet is found in the page, the component automatically attempts to push to its initialized data layer (see [Google Tag Manager Quick Start Guide](https://developers.google.com/tag-manager/quickstart)).
+     */
+    pushToGtmDataLayer: ComponentOptions.buildBooleanOption({
+      postProcessing: (value?: boolean, options?: IAnalyticsOptions) => {
+        const logger = new Logger('Analytics');
```
Comment (samisayegh) [inline src/ui/Analytics/Analytics.ts]: Hmm, usually I pass `this` to the Logger to tell it where we are logging from.

File: src/ui/Analytics/Analytics.ts (hunk 15 above)
Comment (samisayegh) [reply src/ui/Analytics/Analytics.ts]: I share your concerns. In trying to help the user, we are adding complexity. I left a comment above suggesting to simplify by not auto-detecting values.

File: src/ui/Analytics/Analytics.ts [hunk 17]
```diff
@@ -158,7 +162,69 @@ export class Analytics extends Component {
...
+
+    /**
+     * The name of the Google Tag Manager data layer initialized in the page.
+     *
+     * **Note:**
+     * Setting this option is only useful if [`pushToGtmDataLayer`]{@link Analytics.options.pushToGtmDataLayer} is explicitly set to `true`, and the [GTM data layer was renamed](https://developers.google.com/tag-manager/devguide#renaming) in the page.
+     *
+     * By default, if `pushToGtmDatalayer` is left `undefined` and a GTM container snippet is found in the page, the component attempts to extract the data layer name from the snippet. If it fails to do so, or if `pushToGtmDataLayer` is explicitly set to `true`, this option defaults to `dataLayer`.
+     */
+    gtmDataLayerName: ComponentOptions.buildStringOption({
```
Comment (samisayegh) [inline src/ui/Analytics/Analytics.ts]: In line with the above, I think we should remove the default name `dataLayer` and force people to set it if they want to push to Google analytics.

File: src/ui/Analytics/Analytics.ts [hunk 18]
```diff
@@ -463,6 +536,26 @@ export class Analytics extends Component {
     );
   }
 
+  private isGtmDataLayerInitialized(): boolean {
```
Comment (samisayegh) [inline src/ui/Analytics/Analytics.ts]: We can simplify this method and use a getter to give:
```
private get isGtmDataLayerInitialized(): boolean {
    const dataLayerName = this.options.gtmDataLayerName;
    
    if (!dataLayerName) {
         return false;
    }
    
    if (!(<any>window)[dataLayerName]) {
        this.logger.warn(`Cannot push to Google Tag Manager data layer: '${dataLayerName}' is undefined.`);
        return false;
    }

    return true;    
}
```

File: unitTests/ui/AnalyticsTest.ts [hunk 19]
```diff
@@ -158,6 +159,90 @@ export function AnalyticsTest() {
...
+        it('should not push to undefined data layer if pushToGtmDataLayer is true', () => {
+          test = Mock.optionsComponentSetup<Analytics, IAnalyticsOptions>(Analytics, {
+            pushToGtmDataLayer: true,
+            gtmDataLayerName: nonExistingDataLayerName
+          });
+          $$(test.env.root).trigger(AnalyticsEvents.analyticsEventReady, data);
+          expect(typeof (<any>window)[nonExistingDataLayerName]).toBe('undefined');
+        });
+
+        it('should catch error attempting to push to invalid data layer', () => {
```
Comment (fbeaudoincoveo) [inline unitTests/ui/AnalyticsTest.ts]: This one doesn't work (in fact, it always succeeds, even when it shouldn't).

I'm leaving it in the PR so you guys can help me out :)

**Comment by fbeaudoincoveo** [comment]
@olamothe, @samisayegh, @ThibodeauJF 

This is pretty much ready for a "final" review.

After talking with Olivier, Sami and Marya, I decided to remove the "automatic container tag detection" behavior altogether, as it added a lot of complexity in the code and only yielded a slight benefit in return, as the current configuration is fairly trivial (you set data-push-to-gtm-data-layer to true and it should work in most cases; data-gtm-data-layer-name is only for edge cases).

I've done UTs, but I'd like to go further with the ones in the LiveAnalyticsClient (i.e., test that the AnalyticsEventReady handler argument is valid). I'll do that later though; for now I have to get back to writing docs 😁

File: unitTests/ui/AnalyticsTest.ts (hunk 19 above)
Comment (samisayegh) [reply unitTests/ui/AnalyticsTest.ts]: When you say it does not work, I see the test is checking that calling `trigger` does not throw an error. Isn't that what we would expect considering the condition on line 497 in Analytics.ts?

```
if (!(<any>window)[dataLayerName]) {
    this.logger.warn(`Cannot push to Google Tag Manager data layer: '${dataLayerName}' is undefined.`);
    return false;
}
```

Would it be better to expect `expect((<any>window)[defaultDataLayerName][0]).toBe(undefined);` instead?

Comment (fbeaudoincoveo) [reply unitTests/ui/AnalyticsTest.ts]: In this test, I do define the data layer, but overwrite its `push` property so it is no longer a function.

I would expect triggering the AnalyticsEventReady event not to throw in such a case, because the `(<any>window)[dataLayerName].push` call is wrapped in a try-catch.

It does not throw right now... but even if I remove the try-catch and rebuild, the test still passes, so it's apparently invalid.

Comment (samisayegh) [reply unitTests/ui/AnalyticsTest.ts]: I see, I would make sure that
1) the code is calling `handleAnalyticsEventReady` in Analytics.ts
2) That there is no code that is overwriting/resetting what you have stored on  window[dataLayerName].
3) That the window in your test is the same window in the source file.
